resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMasterSource.scala

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.yarn
+
+import com.codahale.metrics.{Gauge, MetricRegistry}
+
+import org.apache.spark.metrics.source.Source
+
+private[spark] class ApplicationMasterSource(yarnAllocator: YarnAllocator) extends Source {
+
+  override val sourceName: String = "applicationMaster"
```

Comment:
In case this is the metrics output:

```
-- Gauges ----------------------------------------------------------------------
applicationMaster.numContainersPendingAllocate
             value = 0
applicationMaster.numExecutorsFailed
             value = 3
applicationMaster.numExecutorsRunning
             value = 9
applicationMaster.numLocalityAwareTasks
             value = 0
applicationMaster.numReleasedContainers
             value = 0
...
```

I would suggest to add application id as a prefix to differentiate between different apps.

Reply:
Ah good catch, I was thinking it automatically added the namespace but it looks like that is only on executor and driver instances.  Perhaps we should just add it as system that will append in the spark.metrics.namespace setting.  for yarn I see the applicationmaster metrics the same as the dag scheduler source, executor allocation manager, etc..  Allowing user to control this makes sense to me.  thoughts?

Reply:
the config spark.metrics.namespace already exists.  see the metrics section in http://spark.apache.org/docs/latest/monitoring.html.  But if you look at the code https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala#L129 its only applied for executor and driver metrics.  I think we should have it apply to the yarn metrics as well.

Reply:
I see. But I think we may not get "spark.app.id" in AM side, instead I think we can get yarn application id, so either we can set this configuration with application id, or directly prepend to the source name.

Reply:
I like the idea to make the metric names more app specific. So I will prepend the app ID to the sourcename. And rerun my test.

Reply:
Ah for the client mode yes there is an order issue with spark.app.id. I'm fine with using the yarn app id since that is essentially what the driver executor use anyway, but I think we should also make it configurable.  I like to see these stay consistent.  If the user can set the driver/executor metrics with spark.metrics.namespace we should allow them to set the yarn ones so that they all could have similar prefix.  Perhaps we add a spark.yarn.metrics.namespace?

application_1530654167152_24008.driver.LiveListenerBus.listenerProcessingTime.org.apache.spark.ExecutorAllocationManager$ExecutorAllocationListener
application_1530654167152_25538.2.executor.recordsRead